lokimq/bt_serialize.h

```diff
@@ -909,5 +922,192 @@ class bt_dict_consumer : private bt_list_consumer {
+/// you technically can create invalid BEncoded strings.  Basically you can add list elements
+/// to a dict and dict elements to a list, and obviously that can end badly.  This may be fixed
+/// at a later time.
+struct fixed_buffer_producer
+{
+    char* begin;
+    char* end;
+
+    char* original_begin;
+
+    fixed_buffer_producer(char* begin, char* end) : begin(begin), end(end), original_begin(begin){
+        if (not (end > begin))
+            throw bt_serialize_length_error("Fixed buffer for serialization end pointer < begin pointer.");
+    }
+
+    // For now, used just for strings.  If we end up serializing large numbers as integers this
+    // will be used for that as well.  As this is designed for strings, will optimize for small
+    // integers and won't work for values > 99999.
+    size_t num_digits(size_t value)
+    {
+        size_t digits = 1;
+
+        if (value > 10) digits++;
+        if (value > 100) digits++;
+        if (value > 1000) digits++;
+        if (value > 10000) digits++;
+
+        return digits;
+    }
+
+    void check_space(size_t len)
+    {
+        if (end <= begin or (size_t)(end - begin) < len)
+            throw  bt_serialize_length_error("Static buffer BEncode buffer overflow.");
+    }
+
+    // MUST check for space before calling this.
+    size_t append_size(size_t size)
+    {
+        size_t digits = 1;
+        char* op = begin;
+
+        *(begin++) = (size % 10) + '0';
+        size /= 10;
+
+        while (size > 0)
+        {
+            *(begin++) = (size % 10) + '0';
+            size /= 10;
+            digits++;
+        }
+
+        char* ed = begin - 1;
+        while (ed > op)
+        {
+            char tmp = *ed;
+            *ed-- = *op;
+            *op++ = tmp;
+        }
+
+        return digits;
+    }
+
+    template <typename T>
+    fixed_buffer_producer& append_integer(T number)
+    {
+        static_assert(std::is_integral<T>::value, "Can't BEncode non-integral type as integer.");
+
+        // integer takes min space 3, e.g. i3e or i0e
+        check_space(3);
+        *(begin++) = 'i';
+
+        char* op = begin;
+
+        // first digit always present, even if value is 0
+        *(begin++) = (number % 10) + '0';
+        number /= 10;
+
+        // will have to profile, but one assumes it is faster to check per-digit than
+        // to calculate number of digits then check.
+        while (number > 0)
+        {
+            check_space(2); // need space for the 'e' to end the number, as well as current digit
+            *(begin++) = (number % 10) + '0';
+            number /= 10;
+        }
+
+        char *ed = begin - 1;
+        while (ed > op)
+        {
+            char tmp = *ed;
+            *ed-- = *op;
+            *op++ = tmp;
+        }
+
+        *(begin++) = 'e';
+
+        return *this;
+    }
+
+    template <typename T>
+    fixed_buffer_producer& append_integer(T number, std::string_view key)
```

Comment:
the parameters are backwards

Reply:
I did this because there is a version which doesn't take a key, but I'm happy to swap the order.